test/pkcs12_format_test.c

```diff
@@ -0,0 +1,918 @@
+        goto err;
+
+    EVP_PKEY_free(pkey);
+    return 1;
+err:
+    EVP_PKEY_free(pkey);
+    return 0;
+}
+
+int generate_secretbag(STACK_OF(PKCS12_SAFEBAG) **bags, int secret_nid, const char *secret,
+                       const PKCS12_ATTR *attrs, int num_attrs)
+{
+    PKCS12_SAFEBAG *bag = NULL;
+
+    TEST_info("Adding secret <%s>", secret);
+
+    bag = PKCS12_add_secret(bags, secret_nid, (const unsigned char *)secret, strlen(secret));
+    if (!TEST_ptr(bag))
+        goto err;
+
+    if (!add_attributes(bag, attrs, num_attrs))
+        goto err;
+
+    return 1;
+err:
+    return 0;
+}
+
+/* Generate the PKCS12 encoding and write to memory bio */
+BIO *generate_p12(STACK_OF(PKCS7) *safes, const PKCS12_ENC *mac, const char *outfile)
+{
+    BIO *mbio = BIO_new(BIO_s_mem());
+    PKCS12 *p12 = PKCS12_add_safes(safes, 0);
+
+    if (!TEST_ptr(p12))
+        goto err;
+    sk_PKCS7_pop_free(safes, PKCS7_free);
+
+    if (mac != NULL) {
+        if (!TEST_true(PKCS12_set_mac(p12, mac->pass, strlen(mac->pass),
+                                      NULL, 0, mac->iter, EVP_get_digestbynid(mac->nid))))
+            goto err;
+    }
+    i2d_PKCS12_bio(mbio, p12);
+
+    /* Can write to file here for debug */
+    //write_p12(p12, outfile);
+
+    PKCS12_free(p12);
+    return mbio;
+err:
+    return NULL;
+}
+
+int write_p12(PKCS12 *p12, const char *outfile)
+{
+    BIO *out = BIO_new_file(outfile, "w");
+    if (out == NULL)
+        goto err;
+
+    i2d_PKCS12_bio(out, p12);
+    BIO_free(out);
+    return 1;
+err:
+    return 0;
+}
+
+PKCS12 *from_bio_p12(BIO *bio, const PKCS12_ENC *mac)
+{
+    PKCS12 *p12 = NULL;
+
+    p12 = d2i_PKCS12_bio(bio, NULL);
+    BIO_free(bio);
+    if (!TEST_ptr(p12))
+        goto err;
+    if (mac == NULL) {
+        if (!TEST_false(PKCS12_mac_present(p12)))
+            goto err;
+    } else {
+        if (!check_p12_mac(p12, mac))
+            goto err;
+    }
+    return p12;
+err:
+    PKCS12_free(p12);
+    return NULL;
+}
+
+PKCS12 *read_p12(const char *infile, const PKCS12_ENC *mac)
+{
+    PKCS12 *p12 = NULL;
+    BIO *in = BIO_new_file(infile, "r");
+
+    if (in == NULL)
+        goto err;
+    p12 = d2i_PKCS12_bio(in, NULL);
+    BIO_free(in);
+    if (!TEST_ptr(p12))
+        goto err;
+    if (mac == NULL) {
+        if (!TEST_false(PKCS12_mac_present(p12)))
+            goto err;
+    } else {
+        if (!check_p12_mac(p12, mac))
+            goto err;
+    }
+    return p12;
+err:
+    PKCS12_free(p12);
+    return NULL;
+}
+
+int check_p12_mac(PKCS12 *p12, const PKCS12_ENC *mac)
+{
+    if (!TEST_true(PKCS12_mac_present(p12))
+        || !TEST_true(PKCS12_verify_mac(p12, mac->pass, strlen(mac->pass))))
+       goto err;
+
+    return 1;
+err:
+    return 0;
+}
+
+static int get_custom_oid(void)
+{
+    static int sec_nid = -1;
+
+    if (sec_nid != -1)
+        return sec_nid;
+    if (!TEST_true(OBJ_create("1.3.5.7.9", "CustomSecretOID", "My custom secret OID")))
```

Comment:
Does this get freed up?

Reply:
Seems to be cleaned up in OPENSSL_cleanup() called at exit. Explicit cleanup with OBJ_cleanup() is deprecated.